keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,376 @@
+
+All of three id's inception is from object creation and it dies with object deletion
+
+## Design Details
+
+### Prerequisite
+
+We need to consider three cases:
+- Case1: Requests from kubectl that creating an object
+- Case2: Requests from kubectl other than creating (e.g. updating, deleting) an object
+- Case3: Requests from controllers
+
+The design below is based on the above three cases
+
+The following picture show our design![design](./overview.png)
+
+
+
+we don't have any modifications in kubectl in this design.
+
+### Design of ID propagation in apiserver
+
+**1. Preprocessing handler (othttp handler)**
+
+1.1 Do othttp's original [Extract](https://pkg.go.dev/go.opentelemetry.io/otel/api/propagators#TraceContext.Extract)() to get [SpanContext](https://pkg.go.dev/go.opentelemetry.io/otel/api/trace#SpanContext)
+
+- For request from kubectl, SpanContext is null,  do [Start](https://github.com/open-telemetry/opentelemetry-go/blob/3a9f5fe15f50a35ad8da5c5396a9ed3bbb82360c/sdk/trace/tracer.go#L38)() to start new SpanContext (new traceid and spanid)
+- For request from controller we can get a valid SpanContext(including traceid and spanid), do [Start](https://github.com/open-telemetry/opentelemetry-go/blob/3a9f5fe15f50a35ad8da5c5396a9ed3bbb82360c/sdk/trace/tracer.go#L38)() to update the SpanContext (new spanid)
+
+1.2 Chain SpanContext and initialtraceid to "r.ctx"
+
+- we use r.ctx to propagate those IDs to next handler
+
+**2. inittrace handler**
+
+Implement a new inittrace handler to propagate the initialtraceid.
+
+2.1 do our new Extract() to get initialtraceid from request header
+
+- For request from kubectl we can't get initialtraceid,  chain a empty initialtraceid to a golang ctx
+- For request from controller we can get initialtraceid, chain the initialtraceid to a golang ctx
+
+2.2 chain above golang cxt to r.ctx(updated in above 1.2)
+-  we use r.ctx to propagate those IDs to next handler
+
+**3. Inject SpanContext and initial-trace to request header sent to webhook**
+
+- call othttp's original [Inject](https://pkg.go.dev/go.opentelemetry.io/otel/api/propagators#TraceContext.Inject)() to inject the SpanContext from r.ctx to header
+- call our new Inject() to inject the initialtraceid from r.ctx to header
+
+### Design of ID propagation (controller)
+
+**1. Extract SpanContext and initialtraceid from annotation of object to golang ctx**
+
+**2. Propagate golang ctx through objects**
+
+When controllers create/update/delete an object A based on another B, we propagate context from B to A. E.g.:
+```
+    ctx = traceutil.WithObject(ctx, objB)
+    err = r.KubeClient.CoreV1().Create(ctx, objA...)
+```
+We do propagation across objects without adding traces to that components.
+
+**3. Inject SpanContext and initial-trace to request header sent to apiserver**
```

Comment:
I don't think this needs to happen in the controller.  I think it just needs to happen in the mutating admission controller.  That way there are fewer changes required in-tree.

I guess the difference is that the initial API call won't have initial-trace-id in baggage, so logging from the API Server will be missing that.  But that would only occur when we are creating the object itself.  Maybe the logging can use trace-id as initial-trace-id if it is empty in that case.